AirCasting/SearchAndFollow/AirBeamMeasurements/AirBeamMeasurementsDownloader.swift

```diff
@@ -0,0 +1,71 @@
+// Created by Lunar on 26/04/2022.
+//
+import SwiftUI
+import Resolver
+
+protocol AirBeamMeasurementsDownloader {
+    func downloadStreams(with sessionID: Int, for airbeam: AirBeamStreamPrefix, completion: @escaping (Result<[MeasurementsDownloaderResultModel], Error>) -> Void) throws
+}
+
+final class AirBeamMeasurementsDownloaderDefault: AirBeamMeasurementsDownloader {
+    @Injected private var urlProvider: URLProvider
+    @Injected private var client: APIClient
+    @Injected private var responseValidator: HTTPResponseValidator
+    private let responseHandler = AuthorizationHTTPResponseHandler()
+    
+    func downloadStreams(with sessionID: Int, for airbeam: AirBeamStreamPrefix, completion: @escaping (Result<[MeasurementsDownloaderResultModel], Error>) -> Void) throws {
+        let decoder = JSONDecoder()
+        let group = DispatchGroup()
+        var combinedResult = [MeasurementsDownloaderResultModel]()
+        
+        let requests = prepareRequests(using: sessionID, sensor: airbeam)
+        
+        for request in requests {
+            group.enter()
+            client.requestTask(for: request) { [responseValidator] result, _ in
+                switch result {
+                case .success(let response):
+                    do {
+                        try responseValidator.validate(response: response.response, data: response.data)
+                        let resultPart = try decoder.decode(MeasurementsDownloaderResultModel.self, from: response.data)
+                        combinedResult.append(resultPart)
+                        group.leave()
+                    } catch {
+                        Log.info("Error when downloading one of the AirBeam streams: \(error)")
+                        group.leave()
+                        completion(.failure(error))
+                    }
+                case .failure(let error):
+                    Log.info("Error when downloading one of the AirBeam streams: \(error)")
+                    group.leave()
+                    completion(.failure(error))
+                }
+            }
+        }
+        group.notify(queue: .main) {
+            Log.info("Downloading all of the streams for AirBeam - completed.")
+            completion(.success(combinedResult))
```

Comment:
problem here is - this `notify` block will get called always. So for the scenario that 2 out of 5 stream downloads will fail the completion will get called twice with `.failure(error)` and after that it will get called with `.success(combinedResult)`. Sorry I didn't pick that before... Usually its done like this:
```Swift
var accumulatedResults : [Result] = []
var allErrors: [Error] = []
...
group.notify(queue: .main) {
 // Here, if allErrors is not empty we can decide what to do next
}
```

This way we won't get multiple `completion` calls